packages/cli/src/auth/index.ts

```diff
@@ -1,86 +1,243 @@
+      server.close();
+      clearTimeout(loginTimeoutHandle);
+      reject(new Error("Timed out waiting"));
+    }, 120000);
+  });
+
+  const loginPromise = new Promise<string>((resolve, reject) => {
+    server = http.createServer(async (req, res) => {
+      function finish(error?: Error) {
+        clearTimeout(loginTimeoutHandle);
+        server.close((closeErr?: Error) => {
+          if (error || closeErr) {
+            reject(error || closeErr);
+          }
+        });
+      }
+      res.setHeader("Access-Control-Allow-Origin", "http://localhost:3000");
+      res.setHeader("Access-Control-Allow-Methods", "GET");
+
+      assert(req.url, "This request doesn't have a URL");
+      const { pathname, query } = url.parse(req.url, true);
+      switch (pathname) {
+        case "/auth/callback": {
+          if (query.token) {
+            const token = Array.isArray(query.token) ? query.token[0] : query.token;
+            const theirState = Array.isArray(query.state)
+              ? query.state[0]
+              : query.state;
+
+            // Check if the payload sent back is the same as the one we sent.
+            if (theirState !== JSON.stringify(loggedIn)) {
+              res.writeHead(400, { "Content-Type": "text/plain" }); // send a 400 response
+              res.end("Unauthorized request, state mismatch", () => {
+                finish(new Error("Unauthorized request, state mismatch"));
+              });
+              reject(
+                new Error(chalk.red("Unauthorized request, state mismatch")),
+              );
+            } else {
+              // Save the token to the config file.
+              const configJson = JSON.parse(fs.readFileSync(credsConfigPath, "utf-8"));
+              fs.writeFileSync(credsConfigPath, JSON.stringify({
+                ...configJson,
+                token
+              }), {
+                encoding: "utf8",
+                mode: 0o600,
+              });
+              res.end(() => {
+                finish();
+              });
+              logger.info(chalk.green(`Successfully logged in.`));
+              resolve(token); // resolve promise with secretKey
+            }
+          } else {
+            res.writeHead(400, { "Content-Type": "text/plain" }); // send a 400 response
+            res.end("No authToken received", () => {
+              finish(new Error("No authToken received"));
+            });
+            reject(new Error(chalk.red("No authToken received")));
+          }
+        }
+      }
     });
 
-    try {
-      await validateKey(response.apiSecretKey);
-    } catch (error) {
-      console.log(error);
-      process.exit(1);
-    }
-    cache.set("api-secret-key", response.apiSecretKey);
-    return response.apiSecretKey;
-  } catch (error) {
-    console.log(error);
+    server.listen(8976);
+  });
+  if (props?.browser) {
+    ora(`Opening a link in your default browser: ${urlToOpen}\n`).start();
```

Comment:
this starts a spinner, do we ever "end" it? why a spinner in the first place?

Reply:
Since this is waiting for the dashboard to respond, but I guess the spinner doesn't make sense on opening the link, I'll move it to a different message.